11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+			smart_refctd_ptr<IGPUComputePipeline> pipelines[3];
+			if(!m_device->createComputePipelines(nullptr, { params, 3 }, pipelines))
+				return logFail("Failed to create Compute Pipelines!\n");
+
+			// Push Constants - only need to specify BDAs here
+			PushConstantData pushConstants;
+			pushConstants.colMajorBufferAddress = m_colMajorBufferAddress;
+			pushConstants.rowMajorBufferAddress = m_rowMajorBufferAddress;
+
+			m_computeCmdBuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+
+			// First Axis FFT
+			m_computeCmdBuf->bindComputePipeline(pipelines[0].get());
+			m_computeCmdBuf->bindDescriptorSets(asset::EPBP_COMPUTE, pipelines[0]->getLayout(), 0, 1, &kernelFFTDescriptorSets[0].get());
+			m_computeCmdBuf->pushConstants(pipelines[0]->getLayout(), IShader::E_SHADER_STAGE::ESS_COMPUTE, 0u, sizeof(pushConstants), &pushConstants);
+			// One workgroup per 2 columns
+			m_computeCmdBuf->dispatch(paddedKerDim.width / 2, 1, 1);
+
+			// Pipeline barrier: wait for first axis FFT before second axis can begin
+			IGPUCommandBuffer::SPipelineBarrierDependencyInfo pipelineBarrierInfo = {};
+			decltype(pipelineBarrierInfo)::buffer_barrier_t bufBarrier = {};
+			pipelineBarrierInfo.bufBarriers = { &bufBarrier, 1u };
+			
+			// First axis FFT writes to colMajorBuffer
+			bufBarrier.range.buffer = m_colMajorBuffer;
+
+			// Wait for first compute write (first axis FFT) before next compute read (second axis FFT)
+			bufBarrier.barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			bufBarrier.barrier.dep.srcAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS;
+			bufBarrier.barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			bufBarrier.barrier.dep.dstAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS;
+
+			m_computeCmdBuf->pipelineBarrier(asset::E_DEPENDENCY_FLAGS(0), pipelineBarrierInfo);
+
+			// Now do second axis FFT - no need to push the same constants again I think
+			m_computeCmdBuf->bindComputePipeline(pipelines[1].get());
+			// Same number of workgroups - this time because we only saved half the rows
+			m_computeCmdBuf->dispatch(paddedKerDim.width / 2, 1, 1);
+
+			// Recycle the pipelineBarrierInfo since it's identical, just change buffer access: Second axis FFT writes to rowMajorBuffer
+			bufBarrier.range.buffer = m_rowMajorBuffer;
+
+			// Also set kernel channel images to GENERAL for writing
+			decltype(pipelineBarrierInfo)::image_barrier_t imgBarriers[CHANNELS];
+			pipelineBarrierInfo.imgBarriers = { imgBarriers, CHANNELS };
+			for (auto i = 0u; i < CHANNELS; i++)
+			{
+				imgBarriers[i].image = m_kernelNormalizedSpectrums[i]->getCreationParameters().image.get();
+				imgBarriers[i].subresourceRange.aspectMask =IImage::EAF_COLOR_BIT;
+				imgBarriers[i].subresourceRange.levelCount = 1u;
+				imgBarriers[i].subresourceRange.layerCount = 1u;
+				imgBarriers[i].barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+				imgBarriers[i].barrier.dep.srcAccessMask = ACCESS_FLAGS::NONE;
```

Comment:
when transitioning away from UNDEFINED you wipe the contents of an image, also nothing has ever used them before, so the source masks should always be NONE